### PR TITLE
total github downloads for all releases + all tags

### PR DESCRIPTION
**Pull request by sagiegurari**
Recreated the original pull request (#526) based on comments and with latest badges repo to avoid conflict issues.

### CODE REVIEW

**Comment by sagiegurari** [comment]
By the way the question from my original pull request remains:
I wasn't 100% sure about github\/downloads consolidation, since the URL pattern is different in the current service and wasn't sure if it will catch shorter urls (with less params).
so there are now 2 services for the github\/downloads pattern but they have different params amount.
The original:
````js
camp.route(/^\/github\/downloads\/([^\/]+)\/([^\/]+)\/([^\/]+)\/([^\/]+)\.(svg|png|gif|jpg|json)$/,
````
The added (by me):
````js
camp.route(/^\/github\/downloads\/([^\/]+)\/([^\/]+)\.(svg|png|gif|jpg|json)$/,
````

Will it work?

**Comment by sagiegurari** [comment]
any updates on this one?

**Comment by espadrine** [comment]
Sorry for the delay.

In the case of conflicts, the first definition wins.

However, I feel we can merge the two definitions into one.

```js
camp.route(/^\/github\/downloads\/([^\/]+)\/([^\/]+)\/([^\/]+)\/([^\/]+)\.(svg|png|gif|jpg|json)$/,
```

That defines four slots: user, repo, the last two can either be:

- tag and asset_name
- "latest" and asset_name
- tag and "total"
- "latest" and "total"
- (now, for the new things from this PR) asset_name and undefined
- "total" and undefined

Whether the last one is undefined determines whether we select by tag or over all tags.

**Comment by sagiegurari** [comment]
I updated the code to look for tag total.
There is only one middleware now (the original) with my additional code in it to check for tag = total to activate the different logic.

**Comment by sagiegurari** [comment]
can you please check it?

**Comment by sagiegurari** [comment]
Any updates on this one?

**Comment by espadrine** [comment]
Sorry, every time I come back to it, I end up taking a large amount of time trying to understand why it does not work. It turns out it was an unrelated issue with my setup.

Having tag *and* asset_name be set to `total` feels weird. In my last comment, I suggested having only one `total`. For instance, it feels like this should work: `/github/downloads/atom/atom/total.svg`. Could you detail the reason you stuck to `total/total`?

**Comment by sagiegurari** [comment]
It has to do with the current pattern used in the current code base which expects specific amount of args.
I asked at the past will the middleware get invoked if less params are provided during a request. If so, I can change, if not, either I write a new middleware which you didn't want (see original pull request), or just get that annoying doule total that I have now which I also not fond of

**Comment by espadrine** [comment]
I think we can make the last parameter optional with this regex: `/^\/github\/downloads\/([^\/]+)\/([^\/]+)\/([^\/]+)(\/[^\/]+)?\.(svg|png|gif|jpg|json)$/`.

**Comment by sagiegurari** [comment]
Right, why I didn't think of it :)
I'll do the changes today.

**Comment by sagiegurari** [comment]
I have modified the code.
If you can check it in your env, it would be great.
Thanks.

**Comment by espadrine** [comment]
There's a couple of things to fix, which I commented on https://github.com/sagiegurari/shields/commit/ec97d7a50694a776e25b223c34a5bfe0b9844ca1.

**Comment by sagiegurari** [comment]
Modified but not as you suggested exactly.
Didn't know the match[x] will be null (I use express mostly, never used Camp)
Anyhow, the asset can be provided, it is the tag that needs to be null.
if no tag, than I pull the info from all releases, but I can still filter based on the requested asset, meaning to pull from all releases the total downloads of a specific asset.
so i changed the url regex as well and moved the optional block one back, again no experience with cargo and not sure if thats ok.

as for the second comment about trailing slash, didn't see any issue in the url i'm building to fetch the info from the github rest api, so i'm not sure there is an issue (that part I did test, just not with cargo server).

File: server.js [hunk 1]
```diff
@@ -2600,15 +2600,24 @@ cache(function(data, match, sendBadge, request) {
...
-  var release_path = tag !== 'latest' ? 'tags/' + match[3] : 'latest';
-  var apiUrl = 'https://api.github.com/repos/' + user + '/' + repo + '/releases/' + release_path;
+  var total = true;
+  if (tag) {
+    total = false;
+  }
+
+  var apiUrl = 'https://api.github.com/repos/' + user + '/' + repo + '/releases';
+  if (!total) {
+    var release_path = tag !== 'latest' ? 'tags/' + tag : 'latest';
```
Comment (espadrine) [inline server.js]: I logged the values of tag and release_path, and tag was `"/latest"`, and so release_path was `"tags//latest"` (which still works, as most servers convert double slashes into single slashes), but not `"latest"`, as the code presumably intends to do.

Either we must `tag = tag.slice(1)` somewhere, or we must compare `tag` to things that have a leading slash (ie, `tag !== '/latest'`, and `'tags' + tag`).

**Comment by sagiegurari** [comment]
added 
````js
  if (tag && (tag.indexOf('/') !== -1)) {
    tag = tag.split('/').join('');
  }
````

File: server.js [hunk 2]
```diff
@@ -2600,15 +2600,29 @@ cache(function(data, match, sendBadge, request) {
...
+
+  var tag = match[3]; //null for all releases
   var asset_name = match[4].toLowerCase(); // eg. total, atom-amd64.deb, atom.x86_64.rpm
   var format = match[5];
-  var release_path = tag !== 'latest' ? 'tags/' + match[3] : 'latest';
-  var apiUrl = 'https://api.github.com/repos/' + user + '/' + repo + '/releases/' + release_path;
+
+  if (tag && (tag.indexOf('/') !== -1)) {
+    tag = tag.split('/').join('');
+  }
```
Comment (espadrine) [inline server.js]: Why can't we replace that `if (…) {…}` by a simple `if (tag) { tag = tag.slice(1); }`?

I was almost tempted to merge and do that change myself, but I may be missing something.

**Comment by sagiegurari** [comment]
I think the 'if' is safer to handle all scenarios. Might be wrong, but I think better to prevent the url issue and not by mistake mess up the tag.
To me it is also a bit more understandable in an already over complicated function.

**Comment by espadrine** [comment]
Thanks!

Given the regular expression, there cannot be more than one slash in `tag`, so it was safe to simply use `tag.slice(1)`.

**Comment by sagiegurari** [comment]
Thanks :+1:

**Comment by sagiegurari** [comment]
when is this going to be deployed and available for use?

**Comment by espadrine** [comment]
Thanks for pinging me! It is deployed now.

**Comment by sagiegurari** [comment]
:+1:  thanks, it works good

**Pull request merged**